### PR TITLE
Servicing index use latest entry

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Fixes https://github.com/aspnet/XRE/issues/1170

Please review @ChengTian @lodejard

### CODE REVIEW

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs [hunk 1]
```diff
@@ -83,22 +90,30 @@ public void Initialize(string folderPath)
...
             if (_entries.TryGetValue(new EntryKey(packageId, packageVersion), out entry))
             {
                 var normalizedAssetPath = assetPath.Replace('\\', '/');
+
+                // Go through the entire list of mappings and use the latest one
                 foreach (var mapping in entry.Mappings)
                 {
                     if (string.Equals(normalizedAssetPath, mapping.AssetPath, StringComparison.OrdinalIgnoreCase))
                     {
-                        replacementPath = Path.Combine(_folderPath, mapping.ReplacementPath);
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: Here is the actual fix

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs [hunk 2]
```diff
@@ -28,54 +28,61 @@ public void Initialize(string folderPath)
...
-                        var parts = line.Split(new[] { '=' }, 2);
-                        if (parts.Length != 2)
+                    if (line.StartsWith("#") || string.IsNullOrWhiteSpace(line))
+                    {
+                        continue;
+                    }
+                    var parts = line.Split(new[] { '=' }, 2);
+                    if (parts.Length != 2)
+                    {
+                        Logger.TraceInformation("[{0}]. Line ({1}): malformed servicing file", GetType().Name, lineNumber);
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: Should the trace information to be `[{0}]: Line ....` ?

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs [hunk 3]
```diff
@@ -28,54 +28,61 @@ public void Initialize(string folderPath)
...
+                        continue;
+                    }
+                    parts[0] = parts[0].Trim();
+                    parts[1] = parts[1].Trim();
+                    if (parts[0].Length == 0 || parts[1].Length == 0)
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: Can we add lines before and after `if-else`?

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: This code was already there :) I only changed the log message and extracted a new method

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs (hunk 1 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: Just an idea. Will the entry mappings be big? If so, would it be faster if just reverse the sequence and return the first hit?

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs (hunk 3 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: Understand

**Comment by troydai** [comment]
I have no comments except beside the trace formatting and potential performance impact change. But they're not essential. Looks good to me.

File: src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs (hunk 1 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/ServicingIndex.cs]: I like the idea. Stay tuned for the change :)

**Comment by victorhurdugaci** [comment]
@troydai pushed a change that includes your suggestion for reverse looking in the match list and 2 more tests

**Comment by ChengTian** [comment]
:shipit:

**Comment by troydai** [comment]
:+1: :shipit:

**Pull request merged**